libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,577 @@
+                                                                       (GDestroyNotify) g_ptr_array_unref);
+  g_autoptr(GPtrArray) autoinstall_flatpaks = g_ptr_array_new_with_free_func (g_object_unref);
+
+  /* Repository checked out, read all files in order and build up a list
+   * of flatpaks to auto-install */
+  autoinstall_d_enumerator = g_file_enumerate_children (directory,
+                                                        G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                                        G_FILE_QUERY_INFO_NONE,
+                                                        cancellable,
+                                                        error);
+
+  if (!autoinstall_d_enumerator)
+    return NULL;
+
+  while (TRUE)
+    {
+      GFile *file;
+      GFileInfo *info;
+      GPtrArray *action_refs = NULL;
+
+      if (!g_file_enumerator_iterate (autoinstall_d_enumerator,
+                                      &info, &file,
+                                      cancellable, error))
+        return NULL;
+
+      if (!file || !info)
+        break;
+
+      action_refs = read_flatpak_ref_actions_from_file (file,
+                                                        cancellable,
+                                                        error);
+
+      if (!action_refs)
+        return NULL;
+
+      g_hash_table_insert (ref_actions_for_files,
+                           g_strdup (g_file_info_get_name (info)),
+                           g_steal_pointer (&action_refs));
+    }
+
+  return g_steal_pointer (&ref_actions_for_files);
+}
+
+static guint
+flatpak_ref_hash (gconstpointer data)
+{
+  return g_str_hash (flatpak_ref_get_name (FLATPAK_REF (data)));
+}
+
+static gboolean
+flatpak_ref_equal (gconstpointer a, gconstpointer b)
+{
+  return g_str_equal (flatpak_ref_get_name (FLATPAK_REF (a)),
+                      flatpak_ref_get_name (FLATPAK_REF (b)));
+}
+
+/* Squash actions on the same ref into the last action on that ref, returning
+ * a pointer array of remote ref actions, ordered by the order key in each
+ * remote ref action. */
+static GPtrArray *
+squash_ref_actions_ptr_array (GPtrArray  *ref_actions)
+{
+  g_autoptr(GHashTable) hash_table = g_hash_table_new_full (flatpak_ref_hash,
+                                                            flatpak_ref_equal,
+                                                            g_object_unref,
+                                                            (GDestroyNotify) flatpak_remote_ref_action_unref);
+  GPtrArray *squashed_ref_actions = NULL;
+  gsize i;
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+
+  for (i = 0; i < ref_actions->len; ++i)
+    {
+      FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
+      g_hash_table_replace (hash_table,
+                            g_object_ref (action->ref),
+                            flatpak_remote_ref_action_ref (action));
```

Comment:
Why not add `action` directly here to the pointer array, instead of using a temporary hash table that's thrown away immediately?

Reply:
This is part of the "squashing" process - we only want the most recent action on the `FlatpakRemoteRef` to be considered (recalling that are hashing on the remote ref's name only).